rust/common/error.rs

```diff
@@ -0,0 +1,134 @@
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ *
+ */
+
+#[derive(Debug)]
+pub struct ErrorMessage {
+    pub code: usize,
+    pub message: String,
+}
+
+pub struct ErrorTemplate {
+    template: &'static str,
+    prefix: &'static str,
+    code: usize,
+    padding_len: usize,
+}
+
+impl ErrorTemplate {
+    pub fn format(&self, args: &[&str]) -> ErrorMessage {
+        let expected_arg_count = self.template.matches("{}").count();
+        assert_eq!(
+            expected_arg_count,
+            args.len(),
+            "Message template `{:?}` takes `{}` args but `{}` were provided",
+            self.template,
+            expected_arg_count,
+            args.len()
+        );
+        let mut buffer = format!("[{}{}{}] ", self.prefix, "0".repeat(self.padding_len), self.code);
+        for (i, fragment) in self.template.split("{}").enumerate() {
+            if i > 0 {
+                buffer += args.get(i - 1).unwrap_or(&"{}");
+            }
+            buffer += fragment;
+        }
+        ErrorMessage { code: self.code, message: buffer }
+    }
+}
+
+const fn num_digits(x: usize) -> usize {
+    if x < 10 {
+        1
+    } else {
+        1 + num_digits(x / 10)
+    }
+}
+
+const fn max(x: usize, y: usize) -> usize {
+    if x > y {
+        x
+    } else {
+        y
+    }
+}
+
+macro_rules! max {
+    ($x:literal) => ($x);
+    ($x:literal, $($xs:literal),*) => (max($x, max!($($xs),*)));
+}
+
+macro_rules! error_messages {
+    {code: $code_pfx:literal, type: $message_pfx:literal, $($error_name:ident = $code:literal: $body:literal),* $(,)?} => {
+        error_messages!($code_pfx, num_digits(max!($($code),*)), $message_pfx, $(($error_name, $code, $body)),*);
+    };
+    ($code_pfx:literal, $code_len:expr, $message_pfx:literal, $(($error_name:ident, $code:literal, $body:literal)),* $(,)?) => {
+        $(
+        pub const $error_name: ErrorTemplate = ErrorTemplate {
+            template: concat!($message_pfx, ": ", $body),
+            prefix: $code_pfx,
+            code: $code,
+            padding_len: $code_len - num_digits($code),
+        };
+        )*
+    };
+}
+
+error_messages! {
+   code: "TQL", type: "TypeQL Error",
+   ILLEGAL_STATE = 1: "Illegal internal state on line {}!",
```

Comment:
Changed the message to at least include the line number, as `Result` does not record a backtrace. Would be good to look into recording stack trace in the ErrorMessage somehow.

Reply:
Agreed, scrapped this error type altogether.